README.md
# aliyun-openapi-core-rust-sdk

[![Crates.io](https://img.shields.io/crates/v/aliyun-openapi-core-rust-sdk)](https://crates.io/crates/aliyun-openapi-core-rust-sdk)
[![Documentation](https://docs.rs/aliyun-openapi-core-rust-sdk/badge.svg)](https://docs.rs/aliyun-openapi-core-rust-sdk)
[![Actions Status](https://github.com/r4ntix/aliyun-openapi-core-rust-sdk/workflows/Continuous%20integration/badge.svg)](https://github.com/r4ntix/aliyun-openapi-core-rust-sdk/actions)
[![Crates.io Download Stats](https://img.shields.io/crates/d/aliyun-openapi-core-rust-sdk)](https://crates.io/crates/aliyun-openapi-core-rust-sdk)

Aliyun OpenAPI POP core SDK for Rust.

## Notes

You must know your `AK`(`accessKeyId/accessKeySecret`), and the aliyun product's `endpoint` and `apiVersion`.

For example, The [ECS OpenAPI](https://help.aliyun.com/document_detail/25490.html), the API version is `2014-05-26`.

And the endpoint list can be found at [here](https://help.aliyun.com/document_detail/25489.html), the center endpoint is ecs.aliyuncs.com. Add http protocol `http` or `https`, should be `https://ecs.aliyuncs.com/`.

## Install

Run the following Cargo command in your project directory:

```shell
cargo add aliyun-openapi-core-rust-sdk
```

Or add the following line to your Cargo.toml:

```toml
aliyun-openapi-core-rust-sdk = "1.1.0"
```

## Usage

The RPC style client:

```rust
use std::collections::HashMap;
use std::env;
use std::error::Error;

use aliyun_openapi_core_rust_sdk::client::rpc::RPClient;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "PascalCase")]
struct Region {
    region_id: String,
    region_endpoint: String,
    local_name: String,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "PascalCase")]
struct RegionList {
    request_id: String,
    regions: HashMap<String, Vec<Region>>,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // create rpc style api client.
    let aliyun_openapi_client = RPClient::new(
        env::var("ACCESS_KEY_ID")?,
        env::var("ACCESS_KEY_SECRET")?,
        "https://ecs.aliyuncs.com/",
    );

    // call `DescribeRegions` with empty queries, return `RegionList`
    let response = aliyun_openapi_client
        .clone()
        .version("2014-05-26")
        .get("DescribeRegions")
        .json::<RegionList>()
        .await?;
    println!("DescribeRegions response: {response:#?}");

    // call `DescribeInstances` with queries, return `String`
    let response = aliyun_openapi_client
        .version("2014-05-26")
        .get("DescribeInstances")
        .query([("RegionId", "cn-hangzhou")])
        .text()
        .await?;
    println!("DescribeInstances response: {response}");

    Ok(())
}
```

The ROA style client:

```rust
use std::collections::HashMap;
use std::env;
use std::error::Error;

use aliyun_openapi_core_rust_sdk::client::roa::ROAClient;
use serde::{Deserialize, Serialize};
use serde_json::json;

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "PascalCase")]
struct TranslateData {
    word_count: String,
    translated: String,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "PascalCase")]
struct Translate {
    request_id: String,
    data: TranslateData,
    code: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // create roa style api client.
    let aliyun_openapi_client = ROAClient::new(
        env::var("ACCESS_KEY_ID")?,
        env::var("ACCESS_KEY_SECRET")?,
        "http://mt.aliyuncs.com",
    );

    // create params.
    let mut params = HashMap::new();
    params.insert("SourceText", "你好");
    params.insert("SourceLanguage", "zh");
    params.insert("TargetLanguage", "en");
    params.insert("FormatType", "text");
    params.insert("Scene", "general");

    // call `Translate` with json params, return `Translate`
    let response = aliyun_openapi_client
        .version("2018-04-08")
        .post("/api/translate/web/general")
        .header([("Content-Type".to_string(), "application/json".to_string())])?
        .body(json!(params).to_string())?
        .json::<Translate>()
        .await?;

    println!("Translate response: {response:#?}");

    Ok(())
}
```

The Log Service(SLS) client:
```rust
use std::env;
use std::error::Error;

use aliyun_openapi_core_rust_sdk::client::log_service::LogServiceClient;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Serialize, Deserialize, Debug)]
struct ProjectList {
    projects: Value,
    count: usize,
    total: usize,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // create log service api client.
    let aliyun_openapi_client = LogServiceClient::new(
        env::var("ACCESS_KEY_ID")?,
        env::var("ACCESS_KEY_SECRET")?,
        "https://cn-hangzhou.log.aliyuncs.com",
    );

    // call `ListProject` api.
    let response = aliyun_openapi_client
        .clone()
        .get("/")
        .json::<ProjectList>()
        .await?;
    println!("ListProject response: {response:#?}");05.
    9+4/154

    Ok(())
}

```

## Examples

Export AK info to env, then run `cargo run --example <NAME>`:

```sh
export ACCESS_KEY_ID=<access_key_id>
export ACCESS_KEY_SECRET=<access_key_secret>

# ecs example
cargo run --example ecs

# rds example
cargo run --example rds

# slb example
cargo run --example slb

# vpc example
cargo run --example vpc

# log service(SLS) example
cargo run --example log_service
```

## License
The MIT License


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.1.0] - 2023-07-15

### Bug Fixes

- Fix link in README.md

### Documentation

- Update log_service example
- Update ros example
- Update vpc example
- Update slb example
- Update sc example
- Update rds example

### Features

- Add log_service client
- Make log service client cloneable

### Miscellaneous Tasks

- Remove useless trait
- Add log service example and update test
- Update log_service example
- Update base16ct and base64 version
- Update author info in LICENSE file
- Add editor config file
- Add rustfmt.toml file
- Use max_width default setting in rustfmt.toml file
- Add log_service example in README.md
- Update RPC & ROA client example
- Add dependabot.yml of github
- Update lib doc
- Update changelog

## [1.0.0] - 2023-06-17

### Bug Fixes

- Fix signature bug with `POST` method in rpc style client
- Fix lib document test error

### Documentation

- Update badges
- Add badges link
- Add NLP example
- Add usage of ROA style client

### Features

- Add roa style client
- Add timeout function for ROA style client
- Add timeout function for RPC style client
- Add `post` function for RPC style client
- Add `post` function for ROA style client
- Update ROA style client
- Add async rpc style client
- Add async roa style client
- Make async client cloneable

### Miscellaneous Tasks

- Add cargo-audit for ci workflow
- Upgrade dependencies
- Make clippy happy
- Update lib document
- Update ROA style client test and lib document
- Update ecs api example
- Update RPC style client examples
- Update ROA style client examples
- Add deprecated message for old client
- Update examples
- Update async RPC style client test
- Update README for release 1.0.0

### Other

- Merge tag 'v0.3.0' into develop

no message

- Merge branch 'hotfix/fix-doc-badges'

- Merge tag 'fix-doc-badges' into develop

no message


### Refactor

- Announcing the new async client
- Move unit tests from tests directory to the same source file
- Remove chrono, rust-crypto dependencies
- Use uuid::v4 for nonce field sign

### Testing

- Add compatibility test case
- Cargo fmt
- Add test case for ROA style client
- Disable deprecated warning of compatibility tests
- Add unit test for async RPC style client

## [0.3.0] - 2019-09-21

### Documentation

- Add crates badge, remove install description
- Add github actions badge

### Miscellaneous Tasks

- Add ci by github actions

### Other

- Merge branch 'release/v0.3.0'


### Refactor

- New function(`get`, `query`, `send`) for RPC style client

### Testing

- Add test case for RPC style client

## [0.2.0] - 2019-09-01

### Refactor

- Pass params of `RPClient::request` from by-copy to by-borrow

## [0.1.0] - 2019-08-31

### Other

- Initial commit


<!-- generated by git-cliff -->


src/lib.rs
//! Aliyun OpenAPI POP core SDK for Rust.
//!
//! ## Notes
//!
//! You must know your `AK`(`accessKeyId/accessKeySecret`), and the aliyun product's `endpoint` and `apiVersion`.
//!
//! For example, The [ECS OpenAPI](https://help.aliyun.com/document_detail/25490.html), the API version is `2014-05-26`.
//!
//! And the endpoint list can be found at [here](https://help.aliyun.com/document_detail/25489.html), the center endpoint is ecs.aliyuncs.com. Add http protocol `http` or `https`, should be `https://ecs.aliyuncs.com/`.
//!
//! ## Install
//!
//! Run the following Cargo command in your project directory:
//!
//! ```shell
//! cargo add aliyun-openapi-core-rust-sdk
//! ```
//!
//! Or add the following line to your Cargo.toml:
//!
//! ```toml
//! aliyun-openapi-core-rust-sdk = "1.0.0"
//! ```
//!
//! ## Usage
//!
//! The RPC style client:
//!
//! ```rust
//! use std::collections::HashMap;
//! use std::env;
//! use std::error::Error;
//!
//! use aliyun_openapi_core_rust_sdk::client::rpc::RPClient;
//! use serde::{Deserialize, Serialize};
//!
//! #[derive(Serialize, Deserialize, Debug)]
//! #[serde(rename_all = "PascalCase")]
//! struct Region {
//!     region_id: String,
//!     region_endpoint: String,
//!     local_name: String,
//! }
//!
//! #[derive(Serialize, Deserialize, Debug)]
//! #[serde(rename_all = "PascalCase")]
//! struct RegionList {
//!     request_id: String,
//!     regions: HashMap<String, Vec<Region>>,
//! }
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn Error>> {
//!     // create rpc style api client.
//!     let aliyun_openapi_client = RPClient::new(
//!         env::var("ACCESS_KEY_ID")?,
//!         env::var("ACCESS_KEY_SECRET")?,
//!         "https://ecs.aliyuncs.com/",
//!     );
//!
//!     // call `DescribeRegions` with empty queries, return `RegionList`
//!     let response = aliyun_openapi_client
//!         .clone()
//!         .version("2014-05-26")
//!         .get("DescribeRegions")
//!         .json::<RegionList>()
//!         .await?;
//!     println!("DescribeRegions response: {response:#?}");
//!
//!     // call `DescribeInstances` with queries, return `String`
//!     let response = aliyun_openapi_client
//!         .version("2014-05-26")
//!         .get("DescribeInstances")
//!         .query([("RegionId", "cn-hangzhou")])
//!         .text()
//!         .await?;
//!     println!("DescribeInstances response: {response}");
//!
//!     Ok(())
//! }
//! ```
//!
//! The ROA style client:
//!
//! ```rust
//! use std::collections::HashMap;
//! use std::env;
//! use std::error::Error;
//!
//! use aliyun_openapi_core_rust_sdk::client::roa::ROAClient;
//! use serde::{Deserialize, Serialize};
//! use serde_json::json;
//!
//! #[derive(Serialize, Deserialize, Debug)]
//! #[serde(rename_all = "PascalCase")]
//! struct TranslateData {
//!     word_count: String,
//!     translated: String,
//! }
//!
//! #[derive(Serialize, Deserialize, Debug)]
//! #[serde(rename_all = "PascalCase")]
//! struct Translate {
//!     request_id: String,
//!     data: TranslateData,
//!     code: String,
//! }
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn Error>> {
//!     // create roa style api client.
//!     let aliyun_openapi_client = ROAClient::new(
//!         env::var("ACCESS_KEY_ID")?,
//!         env::var("ACCESS_KEY_SECRET")?,
//!         "http://mt.aliyuncs.com",
//!     );
//!
//!     // create params.
//!     let mut params = HashMap::new();
//!     params.insert("SourceText", "你好");
//!     params.insert("SourceLanguage", "zh");
//!     params.insert("TargetLanguage", "en");
//!     params.insert("FormatType", "text");
//!     params.insert("Scene", "general");
//!
//!     // call `Translate` with json params, return `Translate`
//!     let response = aliyun_openapi_client
//!         .version("2018-04-08")
//!         .post("/api/translate/web/general")
//!         .header([("Content-Type".to_string(), "application/json".to_string())])?
//!         .body(json!(params).to_string())?
//!         .json::<Translate>()
//!         .await?;
//!
//!     println!("Translate response: {response:#?}");
//!
//!     Ok(())
//! }
//! ```
//!
//! ## Examples
//!
//! Export AK info to env, then run `cargo run --example <NAME>`:
//!
//! ```sh
//! export ACCESS_KEY_ID=<access_key_id>
//! export ACCESS_KEY_SECRET=<access_key_secret>
//!
//! # ecs example
//! cargo run --example ecs
//!
//! # rds example
//! cargo run --example rds
//!
//! # slb example
//! cargo run --example slb
//!
//! # vpc example
//! cargo run --example vpc
//!
//! # log service(SLS) example
//! cargo run --example log_service
//! ```
#![allow(deprecated)]

mod roa;
mod rpc;

#[deprecated(
    since = "1.0.0",
    note = "Please use the `aliyun_openapi_core_rust_sdk::client::roa::ROAClient` instead"
)]
pub use crate::roa::Client as ROAClient;

#[deprecated(
    since = "1.0.0",
    note = "Please use the `aliyun_openapi_core_rust_sdk::client::rpc::RPClient` instead"
)]
pub use crate::rpc::Client as RPClient;

pub mod client;
